### ISSUE TITLE
Lack of Square Function

### ISSUE THREAD

**Issue by jhirshman** [pytorch/pytorch#2618]
Unless, I'm mistaken pytorch lacks a function to square tensors (unlike numpy which has np.square).  There are of course many ways of getting around this.  For a while, I chose to do torch.pow(a, 2).  However, I noticed that this function is really slow.  Upon investigating, I found that it is 30x slower than a * a or torch.mul(a, a).

I'm opening this issue to encourage the inclusion of a square function so that others don't make a similar mistake.

![image](https://user-images.githubusercontent.com/876992/30041327-1ca25c30-919e-11e7-88ec-f8944132ea77.png)

**Comment by fmassa** [comment]
While adding a `square` function is an easy solution, I wonder if for this particular case it wouldn't be better to dispatch the `pow` implementations depending on the power that is used, and would benefit users right away.
This is something that I thought should be done already by cmath, but it seems that it's not the case.
Instead of writing our own version of `TH_pow` and replacing it in [here](https://github.com/pytorch/pytorch/blob/master/torch/lib/TH/generic/THTensorMath.c#L2915), which would incur evaluating conditions for every element of the array, we could just replace [this line](https://github.com/pytorch/pytorch/blob/master/torch/lib/TH/generic/THTensorMath.c#L2915) with a dedicated function that dispatches to `a * a` if `base == 2`.

Something like
```c
void THTensor_(pow)(THTensor *r_, THTensor *t, real value)
{
  THTensor_(resizeAs)(r_, t);
  if(value == 2) {
    THTensor_(cmul)(r_, t_, t_);
  }
  else {
    TH_TENSOR_APPLY2(real, t, real, r_, *r__data = CFUNC(*t_data, value););
  }
}   
```

What do you think?

**Comment by jhirshman** [comment]
That solution would definitely be smart, especially in the short term.  I've found a lot of code that squares via the inefficient power function.

However, I think it would be smart to implement a square function nonetheless to make the difference in behavior more clear.

**Comment by fmassa** [comment]
In the one hand, we have a `sqrt` function, so adding a `square` function makes sense. In the other hand, I'd pretty much prefer avoid bloating the API with "redundant" functions (and we get the same behavior with less typing by just doing `a ** 2`).
But I'll let @soumith decide on that.

**Comment by soumith** [comment]
let's immediately shortcut pow to use cmul for value == 2. i dont think `square` is entirely necessary at this stage. willing to reconsider if enough people demand.

**Comment by fmassa** [comment]
Ok, I'll send a PR soon for TH, not sure if THC needs it as well?

**Comment by soumith** [comment]
looks like it does need the patch as well: https://github.com/pytorch/pytorch/blob/master/torch/lib/THC/generic/THCTensorMathPointwise.cu#L166-L178

**Comment by soumith** [comment]
fixed via #2635
